core/src/sumeragi/main_loop.rs

```diff
@@ -445,11 +445,10 @@ impl Sumeragi {
                     "Block sync update received"
                 );
 
-                // Release block writer before creating new one
-                // FIX: Restore `voting_block` if `handle_block_sync` returns Err
-                // Currently it's not possible because block writer needs to be released
-                // Look at https://github.com/hyperledger/iroha/issues/4643
-                let _ = voting_block.take();
+                if categorize_block_sync(&block, &state.view()).is_ok() {
```

Comment:
maybe we can pass result of `categorize_block_sync` into `handle_block_sync` to avoid doing the same check twice?

in the tests you can define functions which would combine this two function